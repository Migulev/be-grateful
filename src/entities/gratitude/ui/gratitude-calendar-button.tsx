import { forwardRef } from 'react'

import { CalendarDays } from 'lucide-react'

import { Button } from '@/shared/components/ui/button'
import { cn } from '@/shared/utils'

export const GratitudeCalendarButton = forwardRef<
  HTMLButtonElement,
  {
    isActive: boolean
    onClick: () => void
    disabled: boolean
    className?: string
  }
>(({ isActive, onClick, disabled, className }, ref) => {
  return (
    <Button
      onClick={onClick}
      className={cn(className, 'size-10 rounded-b-none p-2 opacity-50', {
        'opacity-100': isActive,
      })}
      disabled={disabled}
      ref={ref}
    >
      <CalendarDays className="size-6" />
    </Button>
  )
})
