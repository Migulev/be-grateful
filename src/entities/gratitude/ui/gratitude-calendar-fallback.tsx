import { Loader } from 'lucide-react'

import { cn } from '@/shared/utils'

export const GratitudeCalendarFallback = ({
  className,
}: {
  className?: string
}) => {
  return (
    <div
      className={cn(
        className,
        'z-10 flex h-[282px] w-[254px] items-center justify-center rounded  border border-primary bg-secondary',
      )}
    >
      <Loader className="animate-spin-slow" />
    </div>
  )
}
