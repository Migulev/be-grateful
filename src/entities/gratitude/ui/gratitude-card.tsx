import { Trash2 } from 'lucide-react'

import {
  Card,
  CardContent,
  CardHeader,
  CardTitle,
} from '@/shared/components/ui/card'
import { BrowserView, MobileView } from '@/shared/libs/device-type'
import { DurationTW } from '@/shared/types'
import { cn, convertDurationTW, formatDate } from '@/shared/utils'

export const GratitudeCard = ({
  date,
  text,
  onDelete,
  isOptimistic = false,
  optimisticDuration = 0,
  isMock = false,
}: {
  date: string
  text: string
  onDelete: () => void
  isOptimistic?: boolean
  optimisticDuration?: DurationTW
  isMock?: boolean
}) => {
  const formattedDate = formatDate(date)

  return (
    <Card
      className={cn(
        // !dev: color hardcode
        `group/trash relative min-h-32 border-pink-200 bg-blue-800/30 transition`,
        `${isOptimistic && 'opacity-50'}`,
        convertDurationTW(optimisticDuration),
      )}
    >
      <CardHeader>
        {/* // !dev: color hardcode */}
        <CardTitle className="font-sans text-xs text-gray-700/80">
          {isMock ? '' : formattedDate}
        </CardTitle>
      </CardHeader>
      <CardContent>
        {/* // !dev: color hardcode */}
        <p className="break-words text-lg text-gray-200/90">{text}</p>
      </CardContent>
      {/* // !dev: color hardcode */}
      {isMock ? null : (
        <BrowserView>
          <div className="group/circle absolute right-4 top-4 flex size-8 items-center justify-center rounded-full transition hover:bg-red-50">
            <Trash2
              onClick={onDelete}
              // !dev: color hardcode
              className="size-5 cursor-pointer text-blue-700 opacity-0 transition group-hover/circle:text-red-700 group-hover/trash:opacity-100"
            />
          </div>
        </BrowserView>
      )}
    </Card>
  )
}
