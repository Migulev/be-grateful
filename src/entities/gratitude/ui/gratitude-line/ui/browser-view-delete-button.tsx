import { EllipsisVertical, X } from 'lucide-react'

import { BrowserView } from '@/shared/libs/device-type'

export const BrowserViewDeleteButton = ({
  onDelete,
}: {
  onDelete: () => void
}) => {
  return (
    <BrowserView>
      <div
        className="group/button absolute -left-5 flex h-full w-10 items-center justify-center"
        role="browser-view-delete-button"
      >
        <X
          className="size-5 cursor-pointer text-destructive opacity-0 group-hover/button:opacity-100"
          onClick={onDelete}
          role="x-delete-button"
        />
        <EllipsisVertical className="absolute size-5 cursor-pointer opacity-0 group-hover/button:invisible group-hover/line:opacity-50" />
      </div>
    </BrowserView>
  )
}
