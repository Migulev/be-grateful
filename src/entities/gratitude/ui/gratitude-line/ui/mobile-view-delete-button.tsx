import { useRef, useState } from 'react'

import { EllipsisVertical, X } from 'lucide-react'

import { MobileView } from '@/shared/libs/device-type'
import { useOutsideClick } from '@/shared/utils'

export const MobileViewDeleteButton = ({
  onDelete,
}: {
  onDelete: () => void
}) => {
  const [isOpenDelete, setIsOpenDelete] = useState(false)
  const refDiv = useRef<SVGSVGElement>(null)
  useOutsideClick(refDiv, () => setIsOpenDelete(false))
  return (
    <MobileView>
      {!isOpenDelete && (
        <EllipsisVertical
          onClick={() => setIsOpenDelete(true)}
          className="absolute -left-3 size-5 cursor-pointer opacity-50"
          role="mobile-view-delete-button"
        />
      )}
      {isOpenDelete && (
        <X
          className="absolute -left-3 size-5 cursor-pointer text-destructive"
          onClick={onDelete}
          ref={refDiv}
          role="x-delete-button"
        />
      )}
    </MobileView>
  )
}
