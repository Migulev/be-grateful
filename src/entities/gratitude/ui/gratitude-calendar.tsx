import { CalendarDays } from 'lucide-react'

import { Button } from '@/shared/components/ui/button'
import { cn } from '@/shared/utils'

export const GratitudeCalendar = ({ className }: { className?: string }) => {
  return (
    <Button className={cn(className, 'rounded-b-none size-10 p-2')}>
      <CalendarDays className="size-6" />
    </Button>
  )
}
