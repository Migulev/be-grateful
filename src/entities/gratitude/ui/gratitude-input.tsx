import { Plus } from 'lucide-react'
import { useRef, useState } from 'react'

import { Button } from '@/shared/components/ui/button'
import { Textarea } from '@/shared/components/ui/textarea'
import { useAuthModal } from '@/shared/libs/modals/auth-modal-context'
import { toastError } from '@/shared/libs/toast'
import { cn, useResizeTextarea } from '@/shared/utils'

import { maxGratitudeTextLength } from '../model/rules'

export const GratitudeInput = ({
  className,
  placeholder,
  onCreateAsync,
  isPending,
  isAuth,
}: {
  className?: string
  placeholder: string
  onCreateAsync: (gratitudeText: string) => Promise<void>
  isPending: boolean
  isAuth: boolean
}) => {
  const [gratitude, setGratitude] = useState('')
  const textareaRef = useRef<HTMLTextAreaElement>(null)
  useResizeTextarea(textareaRef, gratitude)
  const { setIsOpenAuthModal } = useAuthModal()

  const handleInputChange = (e: React.ChangeEvent<HTMLTextAreaElement>) => {
    if (e.target.value.length <= maxGratitudeTextLength) {
      setGratitude(e.target.value)
    } else {
      toastError('300 символов максимум')
    }
  }

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    if (!isAuth) {
      setIsOpenAuthModal(true)
      return
    }
    if (gratitude) {
      await onCreateAsync(gratitude)
      setGratitude('')
    }
  }

  return (
    <form
      onSubmit={handleSubmit}
      className={cn(className, 'flex w-full max-w-xl')}
      onKeyDown={e => {
        if (e.key === 'Enter') handleSubmit(e)
      }}
    >
      <Textarea
        ref={textareaRef}
        onChange={handleInputChange}
        value={gratitude}
        placeholder={placeholder}
        autoFocus
        isInput
      />
      <Button
        type="submit"
        // !dev: color hardcode
        className="ml-2 bg-blue-600 text-white hover:bg-blue-700"
        disabled={isPending}
      >
        <Plus className="h-4" />
      </Button>
    </form>
  )
}
