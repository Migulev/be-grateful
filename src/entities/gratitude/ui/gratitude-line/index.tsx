import { type DurationTW } from '@/shared/global-types'
import { cn, convertDurationTW } from '@/shared/utils'

import { BrowserViewDeleteButton } from './ui/browser-view-delete-button'
import { MobileViewDeleteButton } from './ui/mobile-view-delete-button'

export const GratitudeLine = ({
  title,
  onDelete,
  isOptimistic = false,
  optimisticDuration = 0,
}: {
  title: string
  onDelete: () => void
  isOptimistic?: boolean
  optimisticDuration?: DurationTW
}) => {
  return (
    <div
      className={cn(
        `group/line relative`,
        `${isOptimistic && 'opacity-50'}`,
        convertDurationTW(optimisticDuration),
      )}
    >
      <div className="flex items-center border-b border-muted py-4 pl-8 pr-0 text-lg sm:pr-2">
        <p className="overflow-hidden break-words text-secondary-foreground">
          {`-  ${title}`}
        </p>
        <BrowserViewDeleteButton onDelete={onDelete} />
        <MobileViewDeleteButton onDelete={onDelete} />
      </div>
    </div>
  )
}
