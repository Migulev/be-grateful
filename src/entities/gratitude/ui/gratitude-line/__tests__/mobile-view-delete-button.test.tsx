import { fireEvent, render, screen } from '@testing-library/react'
import { Mock } from 'vitest'

import { useStrictContext } from '@/shared/utils'

import { MobileViewDeleteButton } from '../ui/mobile-view-delete-button'

vitest.mock('@/shared/utils', async importOriginal => {
  const actual = await importOriginal()
  if (typeof actual === 'object' && actual !== null) {
    return {
      ...actual,
      useStrictContext: vitest.fn(),
    }
  }
  return {
    useStrictContext: vitest.fn(),
  }
})

describe('MobileViewDeleteButton', () => {
  test('renders the MobileViewDeleteButton and handles delete action', () => {
    ;(useStrictContext as Mock).mockReturnValue({ isTouchScreen: true })
    const mockOnDelete = vitest.fn()

    render(<MobileViewDeleteButton onDelete={mockOnDelete} />)

    const ellipsisButton = screen.getByRole('mobile-view-delete-button')
    expect(ellipsisButton).toBeInTheDocument()

    fireEvent.click(ellipsisButton)

    const deleteButton = screen.getByRole('x-delete-button')
    expect(deleteButton).toBeInTheDocument()

    fireEvent.click(deleteButton)

    expect(mockOnDelete).toHaveBeenCalledTimes(1)
  })
})
