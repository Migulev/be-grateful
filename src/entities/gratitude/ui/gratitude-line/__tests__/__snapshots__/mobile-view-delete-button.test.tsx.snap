// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MobileViewDeleteButton > renders the MobileViewDeleteButton and handles delete action 1`] = `
<svg
  class="lucide lucide-ellipsis-vertical absolute -left-3 size-5 cursor-pointer opacity-50"
  fill="none"
  height="24"
  role="mobile-view-delete-button"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="2"
  viewBox="0 0 24 24"
  width="24"
  xmlns="http://www.w3.org/2000/svg"
>
  <circle
    cx="12"
    cy="12"
    r="1"
  />
  <circle
    cx="12"
    cy="5"
    r="1"
  />
  <circle
    cx="12"
    cy="19"
    r="1"
  />
</svg>
`;

exports[`MobileViewDeleteButton > renders the MobileViewDeleteButton and handles delete action 2`] = `
<svg
  class="lucide lucide-x absolute -left-3 size-5 cursor-pointer text-destructive"
  fill="none"
  height="24"
  role="x-delete-button"
  stroke="currentColor"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="2"
  viewBox="0 0 24 24"
  width="24"
  xmlns="http://www.w3.org/2000/svg"
>
  <path
    d="M18 6 6 18"
  />
  <path
    d="m6 6 12 12"
  />
</svg>
`;
